#159204721 User Create, Update and Delete Articles

Oluwafayokemi opened the pull request:
#### What does this PR do?
An authenticated user can create, update and delete an article
#### Description of Task to be completed?
- create models for an article, comment, reply, tag and rating
- An authenticated user should be able to create only the articles belonging to him
- An authenticated user should be able to update only the articles belonging to him
- An authenticated user should be able to delete only the articles belonging to him
- Generate two different unique slugs when two users create articles of the same title
- Article slugs should be automatically added to articles if they do not exist
- Add Timestamp to the articles
- Write unit test(s)
#### How should this be manually tested?
- Run npm  run dev or npm start on your terminal
##### Open postman:
- To create an article make a post request to http://localhost:3000/api/articles
- To update an article make a put request to http://localhost:3000/api/articles/:slug
- To delete an article make a delete request to http://localhost:3000/api/articles
#### What are the relevant pivotal tracker stories?
[#159204721](https://www.pivotaltracker.com/story/show/159204721)
#### Screenshots (if appropriate)

![image](https://user-images.githubusercontent.com/33626790/43862331-484b2110-9b51-11e8-8793-fa0dc2418db1.png)

![image](https://user-images.githubusercontent.com/33626790/43862341-5063575a-9b51-11e8-8766-a789d9c9910d.png)
![image](https://user-images.githubusercontent.com/33626790/43862353-5718c6c0-9b51-11e8-9097-b7b599524666.png)

Commit:
create models and association for article, tag, comment, reply and rating

Commit:
create models and association for article, tag, comment, reply and rating

Commit:
create models and association for article, tag, comment, reply and rating

Commit:
create models and association for article, tag, comment, reply and rating

Commit:
create models and association for article, tag, comment, reply and rating

Commit:
create models and association for article, tag, comment, reply and rating

Commit:
create models and association for article, tag, comment, reply and rating

Commit:
create models and association for article, tag, comment, reply and rating

Commit:
add social login id to user table

Commit:
add social login id to user table

Commit:
add social login id to user table

Commit:
add social login id to user table

Commit:
add social login id to user table

Commit:
add social login id to user table

Commit:
add social login id to user table

Commit:
feat(article): Create, Update and Delete Article
- Authenticated user should be able to create an article
- Authenticated user should be able to update an article
- Authenticated user should be able to delete an article
- Unit test for each features
[Delivers #159204721]

Commit:
feat(article): Create, Update and Delete Article
- Authenticated user should be able to create an article
- Authenticated user should be able to update an article
- Authenticated user should be able to delete an article
- Unit test for each features
[Delivers #159204721]

Commit:
feat(article): Create, Update and Delete Article
- Authenticated user should be able to create an article
- Authenticated user should be able to update an article
- Authenticated user should be able to delete an article
- Unit test for each features
[Delivers #159204721]

Commit:
feat(article): Create, Update and Delete Article
- Authenticated user should be able to create an article
- Authenticated user should be able to update an article
- Authenticated user should be able to delete an article
- Unit test for each features
[Delivers #159204721]

Commit:
fix merge conflict on article.js

Commit:
- make slug lower case and stringify date string
- fix hound issue on urinary plus-plus

Commit:
- make slug lower case and stringify date string
- fix hound issue on urinary plus-plus

Commit:
- make slug lower case and stringify date string
- fix hound issue on urinary plus-plus

Commit:
make slug lower case and stringify date string

Commit:
make slug lower case and stringify date string

Commit:
make slug lower case and stringify date string

Commit:
Merge branch 'feature/159204721/user-create-articles' of https://github.com/andela/elven-ah into feature/159204721/user-create-articles

Commit:
fix delete comment

Commit:
fix delete comment

Commit:
fix delete comment

madeofhuman commented on server/controllers/ArticleController.js:
```diff
@@ -1,16 +1,180 @@
+          body: newArticle.body,
+          imageUrl: newArticle.imageUrl,
+          Author: username,
+          createdAt: new Date(newArticle.createdAt).toLocaleString('en-GB', { hour12: true }),
```
🙌🏾🙌🏾🙌🏾Nice thinking here, @Oluwafayokemi!!! I notice, however, that you're returning both `authorId` and `Author`. I feel that is a bit redundant.

madeofhuman reviewed (changes requested):
You did a really good job, Fayoke, I especially love how you converted the time to make it more human-readable in the article response. Do find time, however, to address my concern about the `aruthorId` and `Author` returned.

Furthermore, I noticed you were declaring the foreign key columns in the models. You only declare those in the migration file. The association is what is specified in both the migration file and the model file.

Overall, good work!

Commit:
fix: remove author's id and associations on the model table

Commit:
fix: remove author's id and associations on the model table

Commit:
fix: remove author's id and associations on the model table

Commit:
fix: remove author's id and associations on the model table

Commit:
fix: remove author's id and associations on the model table

Commit:
fix: remove author's id and associations on the model table

Commit:
fix: remove author's id and associations on the model table

Commit:
Merge branch 'feature/159204721/user-create-articles' of https://github.com/andela/elven-ah into feature/159204721/user-create-articles

Commit:
Merge branch 'feature/159204721/user-create-articles' of https://github.com/andela/elven-ah into feature/159204721/user-create-articles

Oluwafayokemi commented:
Thank you for the quick feedback CJ. I have implemented the feedback.

On Wed, Aug 8, 2018 at 10:54 PM CJ Odina <notifications@github.com> wrote:

> *@madeofhuman* requested changes on this pull request.
>
> You did a really good job, Fayoke, I especially love how you converted the
> time to make it more human-readable in the article response. Do find time,
> however, to address my concern about the aruthorId and Author returned.
>
> Furthermore, I noticed you were declaring the foreign key columns in the
> models. You only declare those in the migration file. The association is
> what is specified in both the migration file and the model file.
>
> Overall, good work!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/andela/elven-ah/pull/16#pullrequestreview-144624224>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AgEapiXy4qywDin1cquKdbvUtGVIhieBks5uO14tgaJpZM4V0k8f>
> .
>

Veraclins reviewed (changes requested):
@Oluwafayokemi Nice job you did here. There are just a few areas I feel you might want to look into again.
1. How are you ensuring that only a logged in user is able to perform these operations? I may be wrong but I think you should have protected the article routes with the isloggedIn middleware.
2. I see that you are duplicating a lot of codes in the response you are returning to the user. You could create a method that takes the article created or updated and return the response. This way, you will write the code just once and easily change the structure of the response while maintaining the same response.

Veraclins commented on server/controllers/ArticleController.js:
```diff
@@ -1,16 +1,177 @@
+      imageUrl,
+      categoryId,
+      userId: id,
+      createdAt: Date.now(),
```
@Oluwafayokemi, is there a reason you are adding createdAt and updatedAt in your query? Sequelize already does this. I feel you are manually redoing what is already done authomatically.

Oluwafayokemi replied:
@Veraclins I have tested sequelize date and time. It's not wholly dependent. That is why I am creating mine and passing it as a query string. I was also able to manage the created at and updated at, else sequelize was going to return the same date and time for the both of them.

Veraclins reviewed (commented):
@Oluwafayokemi, is there a reason you are adding createdAt and updatedAt in your query? Sequelize already does this. I feel you are manually redoing what is already done authomatically.

Commit:
Merge branch 'feature/159204721/user-create-articles' of https://github.com/andela/elven-ah into feature/159204721/user-create-articles

Commit:
modify comment and reply models and migrateion

Commit:
modify comment and reply models and migrateion

Commit:
modify comment and reply models and migrateion

Commit:
modify comment and reply models and migrateion

chukwuemekachm reviewed (changes requested):
Well done @Oluwafayokemi on completing this task. However, a few key things to note are as follows.

- Just like @Veraclins  pointed out, you should ensure the endpoints are protected. This prevents unauthenticated and unauthorized users from performing operations on those endpoints.

- On line `95` of `ArticleController.js`, I feel sending the slug to the database is redundant since you are not updating it. I might be wrong and so I would like to learn your reasons for doing that.

- From line `68` through `170` of `ArticleController.js` I noticed that you also allowed the user to perform an update and delete operation on a resource in the database without verifying that the resource in question belonged to them. Please try to add a mechanism which verifies that the user trying to perform these operations owns the resources. This would prevent authors from updating or deleting each other's articles.

- From line `29` through `90` of `ArticleValidation` your error responses are not in line with the team's convention and API specs. The team's convention for an error response should contain an `errors` property in the response. The errors property should be an object which has properties of the illegal request payload's properties as arrays. These arrays would contain all the failing validation for their individual properties. Please kindly resolve these so that all endpoints would have a uniform/consistent error response.

Welldone.

Oluwafayokemi commented:
Thank you for the feedback Chima and Clinton. I actually authenticated the
endpoints. I passed the function as a middleware to the article route.
Please check it.
Clinton, I am not sure of what you mean by duplicating code, but I will
look into it. Chima, the reason why I returned the slug was to prove that
the slug doesn't change after you update, but if the team doesn't want it
returned I will remove it. I will attend to every feedback as fast as I
can. Thank you!

On Thu, Aug 9, 2018 at 3:51 PM Chima Chukwuemeka <notifications@github.com>
wrote:

> *@chukwuemekachm* requested changes on this pull request.
>
> Well done @Oluwafayokemi <https://github.com/Oluwafayokemi> on completing
> this task. However, a few key things to note are as follows.
>
>    -
>
>    Just like @Veraclins <https://github.com/Veraclins> pointed out, you
>    should ensure the endpoints are protected. This prevents unauthenticated
>    and unauthorized users from performing operations on those endpoints.
>    -
>
>    In line 95 of ArticleController.js, I feel sending the slug to the
>    database is redundant since you are not updating it. I might be wrong and
>    so I would like to learn why you did it.
>    -
>
>    From line 68 through 170 of ArticleController.js I noticed that you
>    also allowed the user to perform an update and delete operation on a
>    resource in the database without verifying that the resource in question
>    belonged to them. Please try to add a mechanism which verifies that the
>    user trying to perform these operations owns the resources. This would
>    prevent authors from updating or deleting each other's articles.
>    -
>
>    From line 29 through 90 of ArticleValidation your error responses are
>    not in line with the team's convention and API specs. The team's convention
>    for an error response should contain an errors property in the
>    response. The errors property should be an object which has properties of
>    the illegal request payload's properties as arrays. The arrays would
>    contain all the failing validation for individual properties. Please kindly
>    resolve these so that all endpoints would have a uniform/consistent error
>    response.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/andela/elven-ah/pull/16#pullrequestreview-144879139>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AgEapmXs_Ro-p8NYmw6t5C84NqBz0OLTks5uPEyOgaJpZM4V0k8f>
> .
>

Commit:
- write unit tests
- resturcture validation to follow team's convention
- resturcture index route
- ensure only a user that owns the article can delete and update the article
[Delivers #1592042721]

Commit:
- write unit tests
- resturcture validation to follow team's convention
- resturcture index route
- ensure only a user that owns the article can delete and update the article
[Delivers #1592042721]

Commit:
- write unit tests
- resturcture validation to follow team's convention
- resturcture index route
- ensure only a user that owns the article can delete and update the article
[Delivers #1592042721]

Commit:
- write unit tests
- resturcture validation to follow team's convention
- resturcture index route
- ensure only a user that owns the article can delete and update the article
[Delivers #1592042721]

Commit:
fix: user controller impoprt model

Commit:
fix: user controller impoprt model

Commit:
rebase on user reset password

Commit:
rebase on user reset password

Commit:
Pull branch: fix merge conflict

Commit:
fix: return only createdAt when creating an article

Commit:
fix: return only createdAt when creating an article

Commit:
rebase on login with email and username

Commit:
finish: resolve merge conflict

Commit:
add callback url to process.env

madeofhuman reviewed (approved):
Good job finishing this, @Oluwafayokemi. I know it was not easy, especially with the conflicting and changing models and the merge conflicts from hell. That showed great determination. Keep it up.
